pkg/operator/controller/extension/virtualcluster/reconciler.go

```diff
@@ -0,0 +1,316 @@
+	log := logf.FromContext(ctx)
+
+	extension := &operatorv1alpha1.Extension{}
+	if err := r.RuntimeClient.Get(ctx, request.NamespacedName, extension); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.V(1).Info("Object is gone, stop reconciling")
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, fmt.Errorf("error retrieving object from store: %w", err)
+	}
+
+	gardenList := &operatorv1alpha1.GardenList{}
+	// We limit one result because we expect only a single Garden object to be there.
+	if err := r.RuntimeClient.List(ctx, gardenList, client.Limit(1)); err != nil {
+		return reconcile.Result{}, fmt.Errorf("error retrieving garden object: %w", err)
+	}
+	if len(gardenList.Items) == 0 {
+		// in case a garden resource does not exist or is deleted, update the condition, remove the finalizers and exit early.
+		log.Info("No garden found")
+		conditions := NewVirtualClusterConditions(r.Clock, extension.Status)
+		conditions.virtualClusterReconciled = v1beta1helper.UpdatedConditionWithClock(r.Clock, conditions.virtualClusterReconciled, gardencorev1beta1.ConditionFalse, ConditionNoGardenFound, "no garden found")
+		if err := r.updateExtensionStatus(ctx, log, extension, conditions); err != nil {
+			return reconcile.Result{}, fmt.Errorf("error updating extension status: %w", err)
+		}
+		return reconcile.Result{}, r.removeFinalizers(ctx, log, extension)
+	}
+
+	garden := &gardenList.Items[0]
+	gardenClientSet, err := r.GardenClientMap.GetClient(ctx, keys.ForGarden(garden))
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("error retrieving garden client object: %w", err)
+	}
+
+	if extension.DeletionTimestamp != nil {
+		return reconcile.Result{}, r.delete(ctx, log, gardenClientSet.Client(), extension)
+	}
+	return reconcile.Result{}, r.reconcile(ctx, log, gardenClientSet.Client(), extension)
+}
+
+func (r *Reconciler) updateExtensionStatus(ctx context.Context, log logr.Logger, extension *operatorv1alpha1.Extension, updatedConditions VirtualClusterConditions) error {
+	currentConditions := NewVirtualClusterConditions(r.Clock, extension.Status)
+	if extension.Generation == extension.Status.ObservedGeneration && !v1beta1helper.ConditionsNeedUpdate(currentConditions.ConvertToSlice(), updatedConditions.ConvertToSlice()) {
+		return nil
+	}
+
+	patch := client.MergeFrom(extension.DeepCopy())
+	// Rebuild garden currentConditions to ensure that only the currentConditions with the correct types will be updated, and any other
+	// currentConditions will remain intact
+	extension.Status.Conditions = v1beta1helper.BuildConditions(extension.Status.Conditions, updatedConditions.ConvertToSlice(), currentConditions.ConditionTypes())
+	extension.Status.ObservedGeneration = extension.Generation
+
+	// prevent sending empty patches
+	if data, err := patch.Data(extension); err != nil {
+		return fmt.Errorf("failed getting patch data for Extension %s: %w", extension.Name, err)
+	} else if string(data) == `{}` {
```

Comment:
Nit, backquoting is not needed here.
```suggestion
	} else if string(data) == "{}" {
```